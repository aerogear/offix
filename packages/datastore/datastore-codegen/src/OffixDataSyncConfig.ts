/**
 * Config for OffixDataSync plugin
 */
export interface IOffixDataSyncPluginConfig {
    /**
     * The output dir path.
     * The json schema will be written to ${modelOutputDir}/schema/
     * The datasync config will be written to ${modelOutputDir}/config.ts
     */
    modelOutputDir?: string;
}
