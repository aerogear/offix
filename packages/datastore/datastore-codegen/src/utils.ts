import { ModelDefinition } from "@graphback/core";
import { parseMetadata } from "graphql-metadata";
import {
    isListType,
    getNamedType,
    isCompositeType,
    GraphQLType
} from "graphql";
import { mkdirSync } from "fs";

export const isDataSyncClientModel = (model: ModelDefinition) => {
    return parseMetadata("datasync", model.graphqlType);
};

export const makeDirIfNotExists = (path: string) => {
    try {
        mkdirSync(path);
    } catch (error) {
        // nothing to do here, the directory already exists
    }
};

/**
 * Trys to convert the input graphql type to a ts type.
 * If the input type is composite or "ID",
 * "string" is returned instead to represent ids
 *
 * @param type
 * @returns the ts equivalent of the input graphql type
 */
export const convertToTsType = (type: GraphQLType): string => {
    if (isListType(type)) {
        return `${convertToTsType(getNamedType(type))}[]`;
    }
    if (isCompositeType(type) || type.toString() === "ID" || type.toString() === "GraphbackObjectID") {
        return "string";
    }

    let tsType = type.name.toLowerCase();
    if (tsType === "int" || tsType === "float") {
        tsType = "number";
    } else if (tsType !== "string" && tsType !== "boolean") {
        tsType = "any";
    }

    return tsType;
};
