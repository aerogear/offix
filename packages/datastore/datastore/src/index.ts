export * from "./storage";
export * from "./ModelSchema";
export * from "./DataStore";
export * from "./filters";
export * from "./react";
