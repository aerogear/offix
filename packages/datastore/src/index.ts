export * from "./DataStore";
export * from "./filters";
export * from "./ModelSchema";
export * from "./react";
export * from "./storage";
export * as useDatastoreHooks from "./vue";
