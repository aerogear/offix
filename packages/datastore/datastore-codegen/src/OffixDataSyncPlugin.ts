import { GraphbackPlugin, GraphbackCoreMetadata } from "@graphback/core";
import { writeFileSync } from "fs";
import { IOffixDataSyncPluginConfig } from "./OffixDataSyncConfig";
import { createJsonSchema, createModelType } from "./generate-documents";
import { isDataSyncClientModel, makeDirIfNotExists } from "./utils";
import { validateOffixDataSyncPluginConfig } from "./OffixDataSyncPluginValidator";

export const OFFIX_DATASYNC_PLUGIN_NAME = "OffixDataSyncPlugin";

/**
 * This a graphback plugin which generates:
 * - model json schema
 * - model typings
 * which are required by the offix-datasync for
 * every model annotated with @datasync in schema.
 */
export class OffixDataSyncPlugin extends GraphbackPlugin {
    private pluginConfig: Required<IOffixDataSyncPluginConfig>;

    constructor(config?: IOffixDataSyncPluginConfig) {
        super();
        validateOffixDataSyncPluginConfig(config);
        this.pluginConfig = config as Required<IOffixDataSyncPluginConfig>;
    }

    public createResources(metadata: GraphbackCoreMetadata): void {
        const { modelOutputDir } = this.pluginConfig;
        const documents = this.getDocuments(metadata);

        makeDirIfNotExists(modelOutputDir);
        makeDirIfNotExists(`${modelOutputDir}/schema`);
        writeFileSync(`${modelOutputDir}/schema.json`, JSON.stringify(documents.json, null, 2));
        writeFileSync(`${modelOutputDir}/index.ts`, documents.exports);
        writeFileSync(`${modelOutputDir}/types.ts`, documents.types);
    }

    public getDocuments(metadata: GraphbackCoreMetadata) {
        const models = metadata.getModelDefinitions()
            .filter(model => isDataSyncClientModel(model));

        const modelJsonSchemas = models
            .map(model => (createJsonSchema(model)));
        // concatenate all the json documents
        const jsonSchema = modelJsonSchemas
            .reduce((prev, cur) => ({ ...prev, [cur.name]: cur }), {});
        const modelTypes = models.map(model => createModelType(model)).join("\n");

        // TODO use actual model type instead of any for ModelJsonSchema
        const exports = `import { GeneratedModelSchema } from "offix-datastore";
import jsonSchema from "./schema.json";

export const schema = jsonSchema as GeneratedModelSchema;

export * from "./types";
`;


        return {
            json: jsonSchema,
            types: modelTypes,
            exports
        };
    }

    public getPluginName(): string {
        return OFFIX_DATASYNC_PLUGIN_NAME;
    }
}
