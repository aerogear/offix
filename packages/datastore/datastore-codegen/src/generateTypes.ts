import { ModelDefinition } from "@graphback/core";
import { GraphQLOutputType, isNonNullType, getNullableType, isCompositeType } from "graphql";
import { convertToTsType } from "./utils";

const isPrimaryKey = (fieldName: string, type: GraphQLOutputType) => {
    if (isNonNullType(type)) {
        type = getNullableType(type);
    }

    return (type.toString() === "ID");
}

const getPrimaryKey = (model: ModelDefinition) => {
    const fieldMap = model.graphqlType.getFields();

    return Object.keys(fieldMap)
        .find((fieldName) => (isPrimaryKey(fieldName, fieldMap[fieldName].type)))
}

const getField = (fieldName: string, type: GraphQLOutputType) => {
    if (isNonNullType(type)) {
        type = getNullableType(type);
    } else {
        fieldName = `${fieldName}?`;
    }

    return `${fieldName}: ${convertToTsType(type)}`;
}

const getModelProperties = (model: ModelDefinition) => {
    const fieldMap = model.graphqlType.getFields();

    return Object.keys(fieldMap)
        .map(fieldName => (getField(fieldName, fieldMap[fieldName].type)))
        .join(";\n    ");
};

export const createModelType = (model: ModelDefinition) => {
    const modelName = model.graphqlType.name;
    const primaryKey = getPrimaryKey(model);

    return `export interface ${modelName} {
    ${getModelProperties(model)}
}

export type ${modelName}Create = ${primaryKey ? `Omit<${modelName}, "${primaryKey}">` : modelName};
export type ${modelName}Change =  ${primaryKey ? `Pick<${modelName}, "id"> && ` : ""}Partial<${modelName}Create>;
`;
}
