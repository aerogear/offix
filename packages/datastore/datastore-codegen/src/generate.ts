import { GraphbackPluginEngine } from "@graphback/core";
import { readFileSync } from "fs";
import { OffixDataSyncPlugin } from "./OffixDataSyncPlugin";

export const generate = (schemaPath: string, outputPath: string) => {
    const schema = readFileSync(schemaPath).toString();
    const pluginEngine = new GraphbackPluginEngine({
        schema,
        plugins: [
            new OffixDataSyncPlugin({ modelOutputDir: outputPath })
        ]
    });
    pluginEngine.createResources();
};
