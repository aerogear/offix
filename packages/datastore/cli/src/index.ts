import yargs from "yargs";
import { generate } from "./generate";

export { generate };
export { OffixDataSyncPlugin as Plugin } from "./OffixDataSyncPlugin";

export const cmd = () => {
  // eslint-disable-next-line no-unused-expressions
  yargs
    .command(
      "generate [schema] [outputPath]",
      "Generate Datastore config",
      (y) => {
        y.positional("schema", {
          describe: "The schema path",
          default: "./src/models/mock.graphql",
          type: "string"
        }).positional("outputPath", {
          describe: "The output dir path",
          default: "./src/datasync/generated",
          type: "string"
        });
      },
      (argv) => generate(argv.schema as string, argv.outputPath as string)
    )
    .demandCommand(1)
    .help()
    .argv;
};
