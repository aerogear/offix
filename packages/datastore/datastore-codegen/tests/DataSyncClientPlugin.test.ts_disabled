import { readFileSync } from 'fs';
import { GraphbackCoreMetadata } from '@graphback/core';
import { buildSchema } from 'graphql';
import { Plugin } from '../src'

const schemaText = readFileSync(`${__dirname}/mock.graphql`, 'utf8')

test('Test json schema', () => {
  const crudMethods = {
    "create": true,
    "update": true,
    "findOne": true,
    "find": true,
    "delete": true,
  }

  const metadata = new GraphbackCoreMetadata({ crudMethods }, buildSchema(schemaText))
  const plugin = new Plugin({ modelOutputDir: './tmp/output/' });
  expect(plugin.getDocuments(metadata)).toMatchSnapshot();
});

test('Test datasync config', () => {
  const crudMethods = {
    "create": true,
    "update": true,
    "findOne": true,
    "find": true,
    "delete": true,
  }

  const metadata = new GraphbackCoreMetadata({ crudMethods }, buildSchema(schemaText))
  const plugin = new Plugin({ modelOutputDir: './tmp/output/' });
  expect(plugin.getDataSyncConfig(metadata)).toMatchSnapshot();
});
