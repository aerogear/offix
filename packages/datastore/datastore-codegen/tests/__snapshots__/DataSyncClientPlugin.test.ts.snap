// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test datasync config 1`] = `
"import { DataStore } from 'offix-datastore';
import { schema } from './schema';

export const datastore = new DataStore({
    dbName: \\"offix-datasync\\",
    clientConfig: {
      url: \\"http://localhost:4000/graphql\\",
      wsUrl: \\"ws://localhost:4000/graphql\\",
    }
});

export const NoteModel = datastore.setupModel(schema.Note);
export const CommentModel = datastore.setupModel(schema.Comment);

datastore.init();
"
`;

exports[`Test json schema 1`] = `
Object {
  "json": Object {
    "Comment": Object {
      "name": "Comment",
      "primaryKey": "id",
      "properties": Object {
        "description": Object {
          "isRequired": true,
          "key": "description",
          "type": "string",
        },
        "id": Object {
          "index": true,
          "isRequired": true,
          "key": "id",
          "primary": true,
          "type": "string",
        },
        "noteComment": Object {
          "isRequired": true,
          "key": "noteComment",
          "type": "string",
        },
        "title": Object {
          "isRequired": true,
          "key": "title",
          "type": "string",
        },
      },
      "type": "object",
      "version": 1,
    },
    "Note": Object {
      "name": "Note",
      "primaryKey": "id",
      "properties": Object {
        "comments": Object {
          "isRequired": true,
          "key": "comments",
          "type": "string[]",
        },
        "description": Object {
          "isRequired": true,
          "key": "description",
          "type": "string",
        },
        "id": Object {
          "index": true,
          "isRequired": true,
          "key": "id",
          "primary": true,
          "type": "string",
        },
        "title": Object {
          "isRequired": true,
          "key": "title",
          "type": "string",
        },
      },
      "type": "object",
      "version": 1,
    },
  },
  "schemaExport": "import { DataSyncJsonSchema } from \\"offix-datastore\\";
import jsonSchema from \\"./schema.json\\";

type Schema<T = any> = {
    [P in keyof T]: DataSyncJsonSchema<any>
};

export const schema = jsonSchema as Schema;
",
}
`;
